LoRaEngine/modules/LoRaWanNetworkSrvModule/LoRaWan.NetworkServer/LoRaDeviceTelemetry.cs

```diff
@@ -85,29 +77,27 @@ public LoRaDeviceTelemetry()
         {
         }
 
-        public LoRaDeviceTelemetry(Rxpk rxpk, LoRaPayloadData upstreamPayload, object payloadData, byte[] decryptedPayloadData)
+        public LoRaDeviceTelemetry(LoRaRequest request, LoRaPayloadData upstreamPayload, object payloadData, byte[] decryptedPayloadData)
         {
-            if (rxpk is null) throw new ArgumentNullException(nameof(rxpk));
+            if (request is null) throw new ArgumentNullException(nameof(request));
+            var radioMetadata = request.RadioMetadata;
+            if (radioMetadata is null) throw new ArgumentException(nameof(radioMetadata));
             if (upstreamPayload is null) throw new ArgumentNullException(nameof(upstreamPayload));
-            if (rxpk.ExtraData != null)
-                ExtraData = new Dictionary<string, object>(rxpk.ExtraData);
-            Chan = rxpk.Chan;
-            Codr = rxpk.Codr;
+            var datr = request.Region.GetDatarateFromIndex(request.RadioMetadata.DataRate);
             Data = payloadData;
             Rawdata = decryptedPayloadData?.Length > 0 ? Convert.ToBase64String(decryptedPayloadData) : string.Empty;
-            Datr = rxpk.Datr;
-            Freq = rxpk.Freq;
-            Lsnr = rxpk.Lsnr;
-            Modu = rxpk.Modu;
-            Rfch = rxpk.Rfch;
-            Rssi = rxpk.Rssi;
-            Size = rxpk.Size;
-            Stat = rxpk.Stat;
-            Time = rxpk.Time;
-            Tmms = rxpk.Tmms;
-            Tmst = rxpk.Tmst;
             Fcnt = upstreamPayload.GetFcnt();
             Port = upstreamPayload.Fport.Value;
+            Freq = radioMetadata.Frequency.InMega;
+            Datr = datr.ToString();
+            Rssi = radioMetadata.UpInfo.ReceivedSignalStrengthIndication;
+            Rfch = radioMetadata.UpInfo.AntennaPreference;
+            Lsnr = radioMetadata.UpInfo.SignalNoiseRatio;
+            Time = radioMetadata.UpInfo.Xtime.ToString(CultureInfo.InvariantCulture); // This field was just used for telemetry in rxpk. Now it's being used for bringing unaltered the Xtime.
```

Comment:
Why convert to a string? Why not keep the original `ulong` type?

Reply:
I will convert this in ulong